Add output checking statements (#199)
sa-tre/satre-specification: docs/source/pillars/data_management.md

    Results will be produced which address the project's purpose, some of which may be outputs that are removed from the TRE.
    Understanding what these outputs are likely to be and their sensitivity as early as possible will help prepare for their processing and publication.
  - Recommended

> * - 3.3.3.
>   - You must have a process for output checking and assign responsibility for output checking.
>   - Regardless of whether or how outputs are classified and whether the outputs were decided at the project outset, all outputs must be checked according to your process by the responsible party.
>   Output checking ensures that no unintended data breach occurs during the egress procedure.
>   - Mandatory

Reviewer: I've broken this into two pieces and updated the wording. Thoughts @sa-tre/spec-maintainers ?

```suggestion
* - 3.3.3.
  - You must have a process for ensuring that outputs are safe.
  - This process should define what risks you are worried about and how to mitigate them.
     You might choose to follow existing guidelines, for example around statistical disclosure.
  - Mandatory
* - 3.3.4.
  - You must have a process for assigning responsibility for output checking.
  - A named person or team should be given responsibility for checking outputs. This person or team must follow your output checking process.
    Output checking can help mitigate against unintentional data breaches during the egress procedure.
  - Mandatory
```

    Results will be produced which address the project's purpose, some of which may be outputs that are removed from the TRE.
    Understanding what these outputs are likely to be and their sensitivity as early as possible will help prepare for their processing and publication.
  - Recommended
* - 3.3.3.
  - You must have a process for disclosure control that ensures outputs are safe.
  - This process should define expected risks and how to mitigate them.
    All TRE outputs must be subject to this process.

>     You might choose to follow existing guidelines, for example around statistical disclosure.

Reviewer 2: I would expect "for example" to be followed by a specific example.
Is there one we want to draw attention to?

Otherwise, maybe we can just say "You might choose to follow existing guidelines"?

Reviewer 3: With ML models you cannot guarantee outputs are safe so need to change the wording.

Reviewer 3: How about? "You must have a documented process for disclosure control of output files from the TRE."

Author: > With ML models you cannot guarantee outputs are safe

@drchriscole do you think we also need a statement regarding outputs that cannot be easily determined as safe, such as ML models

Something like:

- Statement: You must have a documented policy for handling disclosure risks associated with outputs that cannot be manually checked
- Guidance: It is generally not recommended to egress TRE outputs that cannot be manually checked, but in cases such as when you want to egress a machine learning model, there should be an institutional policy about what to do if such a model leaks data that was kept secure whilst it was inside the TRE
- Mandatory

Reviewer 4: How about

Guidance: It is generally recommended that only files that can be checked are egressed. For example, it is difficult to check for the presence of sensitive data in binary files or very large numeric files. If egress is allowed the risks of inadvertent disclosure must be mitigated and documented.

Author: See 3.3.5 just added

  - Mandatory
* - 3.3.4.
  - You must have a process for assigning responsibility for output checking.

>   - A named person or team should be given responsibility for checking outputs. This person or team must follow your disclosure control process.

Reviewer 2: ```suggestion
  - A named person or team should be given responsibility for checking outputs.
    This person or team must follow your disclosure control process.
```

Reviewer 3: The SACRO project is developing semi-automated tools for output checking so it would be good to make reference to that as well. The responsible person will be responsible for automated checks as well.

    Results will be produced which address the project's purpose, some of which may be outputs that are removed from the TRE.
    Understanding what these outputs are likely to be and their sensitivity as early as possible will help prepare for their processing and publication.
  - Recommended
* - 3.3.3.

>   - You must have a documented process for disclosure control of output files from the TRE.

Reviewer 2: ```suggestion
  - You must have a documented process for disclosure control of outputs from the TRE.
```

We should be careful, I don't think we can assume or guarantee that outputs are always files.
I could, for example, imagine a TRE which produces outputs through some private or public API.

  - This process should define expected risks and how to mitigate them.
    All TRE outputs must be subject to this process.
    You might choose to follow existing guidelines, for example around statistical disclosure.
  - Mandatory
* - 3.3.4.
  - You must have a process for assigning responsibility for output checking.

>   - A named [INSERT OUTPUT CHECKING ROLE] person or team should be given responsibility for checking outputs.

Reviewer: Can you add this to the `roles.md` page and make this a link?

>     This person or team must follow your disclosure control process and will be responsible for any automated parts of this process.

Reviewer 2: What is important here. Is it that we know who does the output checking or that there are people able to do it?

I think it would be better to introduce an output checker role and refer to that here.

Author: We still haven't filled out roles: https://satre-specification.readthedocs.io/en/latest/roles.html

Perhaps now is a good time to have a look at this? Happy to have a first go

Reviewer: Talk to @machintim also as he's done a bit of work here.

Reviewer 4: To avoid blocking PRs could we either:
- remove this for now
- or keep it, and replace it with a link to a role in a follow-up?

Reviewer: Or just link to a role called `Output Checker` that doesn't have any content yet?

>   - Named {ref}`output checker(s) <admin_roles>` should be given responsibility for checking outputs.
>     This person or team must follow your disclosure control process and will be responsible for any automated parts of this process.

Reviewer 2: ```suggestion
  - {ref}`Output checkers <admin_roles>` should be given responsibility for checking outputs.
    They must follow your disclosure control process and will be responsible for any automated parts of this process.
```

>     Output checking can help mitigate against unintentional data breaches during the egress procedure.

Reviewer 2: ```suggestion
    Output checking can help mitigate against unintentional data disclosure or leaks.
```

Breach implies an unauthorised person has (forcibly) gained access to data they shouldn't. What we are talking about here is authorised people accidentally disclosing sensitive data.

  - Mandatory
* - 3.3.5.

>   - Where the possibility exists that a project will require egress of outputs that cannot be manually checked, you must have a documented policy for handling disclosure risks associated with these outputs.

Reviewer 2: ```suggestion
  - You must have a documented policy for handling disclosure risks associated with outputs that cannot be manually checked.
```

Author: Is this too much for mandatory though? Could be wrong but I don't know if this would apply to all, or even most TREs - especially since it's arguably just bad practice

Author: But I think it should be mandatory conditional on the TRE operators/projects wanting to do it

Reviewer 4: I think it can be mandatory- your policy can be that "outputs that can't be manually checked are always refused". If that's not clear then might be worgth adding?

Reviewer: Added `any` to the suggested change.

Author: re-reading the guidance
> It is generally recommended that only files that can be checked are disclosed
I think that covers having a policy like "outputs that can't be manually checked are always refused"

>   - It is generally recommended that only files that can be checked are egressed.

Reviewer 2: ```suggestion
  - It is generally recommended that only files that can be checked are disclosed.
```

>   - It is generally recommended that only files that can be checked are disclosed.
>     For example, it is difficult to check for the presence of sensitive data in binary files or very large numeric files.
>     If egress is allowed, the risks of inadvertent disclosure must be mitigated and documented.

Reviewer: ```suggestion
  - Some categories of output, for instance binary files or very large numeric files, can be difficult to manually check.
    If egress of such files is permitted then the risks of inadvertent disclosure must be mitigated and documented.
    Refusing to allow egress of such files is also a valid policy decision.
```

Author: Nice, reads better
